distributed/worker_state_machine.py

```diff
@@ -231,20 +232,21 @@ class TaskState:
     #: True if the task is in memory or erred; False otherwise
     done: bool = False
 
+    _instances: ClassVar[weakref.WeakSet[TaskState]] = weakref.WeakSet()
+
     # Support for weakrefs to a class with __slots__
     __weakref__: Any = field(init=False)
 
+    def __post_init__(self):
+        TaskState._instances.add(self)
+
     def __repr__(self) -> str:
         return f"<TaskState {self.key!r} {self.state}>"
 
     def __eq__(self, other: object) -> bool:
         if not isinstance(other, TaskState) or other.key != self.key:
             return False
-        # When a task transitions to forgotten and exits Worker.tasks, it should be
-        # immediately dereferenced. If the same task is recreated later on on the
-        # worker, we should not have to deal with its previous incarnation lingering.
-        assert other is self
```

Comment:
@crusaderky: This assertion prevented me from being able to reliably add `TaskState` objects to a `WeakSet` ([https://github.com/hendrikmakait/distributed/runs/6779613623?check_suite_focus=true#step:11:76747](https://github.com/hendrikmakait/distributed/runs/6779613623?check_suite_focus=true#step:11:76747)). Is there a place where we should this on the _user_ side of the `__eq__` check?

Reply:
I'm OK with this change. As long as we cannot reliably dereference TaskState objects we cannot assert this condition. It's also very likely that we'll always require an explicit `gc.collect` to get rid of stale TaskState objects such that we cannot keep this assert.

We may consider adding something like

```python
from itertools import groupby
for key, tss in groupby(TaskState._instances, key=lambda ts: ts.key):
    assert sum(ts.state != "forgotten" for ts in tss) <= 1
```

to `Worker.validate_state`

Reply:
I actually think 
```python
def __eq__(self, other):
    return other is self
```
should be sufficient

Reply:
@fjetter: Agreed

Reply:
@crusaderky: Great suggestion, done.

Reply:
https://github.com/dask/distributed/pull/6525/files#r892142619

Would never work. We often have multiple workers per process, i.e. a task may actually exist in various states.

Reply:
This does in fact cause a number of tests to fail.

Reply:
happy with other is self; however can we have a bit of blurb for posterity please?
```python
def __eq__(self, other):
    # A task may be forgotten and a new TaskState object with the same key may be created in
    # its place later on. In the Worker state, you should never have multiple TaskState objects with
    # the same key. We can't assert it here however, as this comparison is also used in WeakSets
    # for instance tracking purposes.
    return other is self

Reply:
Absolutely, done!